### PR TITLE
Add code infrastructure for pipeline configuration regression tests

### PR DESCRIPTION
**Pull request by nwiltsie**
Okay, this is _a lot_. None of this is in Action format yet, but I wanted to push up this minimal working version and start getting some feedback on it.

# The problem this is solving
We have so much metaprogramming craziness inside the pipeline configurations. The configuration changes based on available CPUs and memory, there are [7 ways to modify parameters](https://www.nextflow.io/docs/latest/config.html#configuration-file), and we have fancy process parameter merging and retry specification.

Given that, it's very difficult to review most pull requests changing configuration logic, as I have exceptionally little understanding of what's going on under the hood.

I want an Action that can run regression tests on our Nextflow pipeline configurations. That is, given a specific commit of a pipeline and a defined configuration (split across config files, parameter files, environment variables, and command line arguments), I want to be able to compute the final configuration _in the cloud_, without requiring any files outside of the repository. Then that Action can automatically run on pull requests, documenting how the code changes impacted the final rendered configuration.

This PR is the bones of that Action (and a record of my slow descent into madness).

# Interface

## Test File Format
Tests are defined as specially-formatted JSON files tracked within each pipeline repository, one test per file. For demonstration I've included one test for the recalibrate BAM pipeline with this PR. The configuration appears at the start:

https://github.com/uclahs-cds/tool-Nextflow-action/blob/c8904845ad2d0be12c7544781d8fe65397c791af/run-nextflow-tests/recalibrate-bam.json#L1-L37

The following keys are required (unless stated otherwise all paths are relative to the pipeline root):

| Key | Description |
| --- | --- |
| config | A list of configuration files to be included (`nextflow -c <file1> -c <file2>`) |
| params_file | A single parameter file or an empty string (`nextflow -params-file <file>`) |
| cpus | The integer CPU count to be returned by `SysHelper::getAvailCpus()` |
| memory_gb | The memory value to be returned by `SysHelper::getAvailMemory` (float, GB) |
| empty_files | A list of absolute paths to `touch` within the docker image |
| mapped_files | A map of relative repository files to absolute container paths that should be mapped into the docker image |
| nf_params | A map of command-line parameters to pass to Nextflow (`nextflow --<key>=<value>`) |
| envvars | A map of environment variables to set in the docker image (`KEY=VALUE nextflow ...`) |
| mocks | Method names to be mocked, mapped to the objects they should return |
| dated_field | A list of JSONPath-like keys indicating values in the rendered configuration that contain datestamps |
| expected_results | The expected output of the test (lightly filtered, see later sections for details) |

That `expected_results` field looks something like this:

https://github.com/uclahs-cds/tool-Nextflow-action/blob/c8904845ad2d0be12c7544781d8fe65397c791af/run-nextflow-tests/recalibrate-bam.json#L38-L810

Due to our metaprogramming madness many of the `memory` and `cpu` values under `process` are closures that return different results depending upon `task.attempt` - in those cases the reported value is a map containing the closure code itself (heavy caveats) and the computed values for attempts 1, 2, and 3:

https://github.com/uclahs-cds/tool-Nextflow-action/blob/c8904845ad2d0be12c7544781d8fe65397c791af/run-nextflow-tests/recalibrate-bam.json#L605-L612

## Current Usage
The `entry.py` script compares the bundled tests against a local checkout of the recalibrate BAM pipeline. With a clean checkout that goes well (all examples are running on my mac laptop, detached from the cluster):

```console
$ ./entry.py ~/src/pipeline-recalibrate-BAM/
No changes!
```

After I make a small change to the pipeline...

```diff
--- a/config/F16.config
+++ b/config/F16.config
@@ -17,7 +17,7 @@ process {
         retry_strategy {
             memory {
                 strategy = 'exponential'
-                operand = 2
+                operand = 4
             }
         }
     }
```

```console
$ ./entry.py ~/src/pipeline-recalibrate-BAM/
.params.proc_resource_params.run_RealignerTargetCreator_GATK.retry_strategy.memory.operand
2
4
------
.process.withName:run_RealignerTargetCreator_GATK.memory.2
8 GB
16 GB
------
.process.withName:run_RealignerTargetCreator_GATK.memory.3
16 GB
31 GB
------
.process.withName:run_RealignerTargetCreator_GATK.memory.closure
retry_updater(4 GB, exponential, 2, $task.attempt, memory)
retry_updater(4 GB, exponential, 4, $task.attempt, memory)
------
Saving updated file to /Users/nwiltsie/src/tool-Nextflow-action/run-nextflow-tests/recalibrate-bam-out.json
```

When differences are found, the script saves a complete test file as "[testfile]-out.json", which makes it easy to copy over the original if the changes were intended.

# Internals

The key steps are:
1. Build the docker image (currently done on-the-fly)
2. Generate a temporary config file with the appropriate mocks and `includeConfig`s
3. Within the image, run a groovy script that's kinda-sorta like `nextflow config -properties`
4. Outside the image, parse the properties and diff them against the expected values

## Docker Image

This tool runs in a modified version of the [nextflow/nextflow:23.10.0](https://hub.docker.com/layers/nextflow/nextflow/23.10.0/images/sha256-f170118647183601452d007311618960aa3b2334689529d0e0f98c102b31458f) image, and can be updated if desired.

The key changes are to download and include [mockito](https://github.com/mockito/mockito) on the classpath, and to swap the JAR entrypoint for the `nextflow` script from `nextflow.cli.Launcher` to `groovy.ui.GroovyMain` (what you get if you call `groovy` from the command line).

This is definitely hackish, but the goal is to execute an arbitrary groovy script with _exactly_ the same JVM arguments and settings as `nextflow`. The `nextflow` script does some tricky business, caching all of those arguments in `/.nextflow/tmp/launcher/nextflow-one_${NEXTFLOW_VERSION}/buildkitsandbox/classpath-${NEXTFLOW_MD5}`, where `${NEXTFLOW_MD5}` is the hash of [a bunch of environment variables](https://github.com/nextflow-io/nextflow/blob/5e2ce9ed82ccbc70ec24a83e04f24b8d45855a78/nextflow#L377-L399). I'd prefer to use [`$NXF_CLASSPATH`](https://www.nextflow.io/docs/latest/config.html) rather than `sed` to splice in the extra JARs, but doing so would result in an unpredictable launcher and make it harder to swap out the entrypoint.

## Runtime Container

At runtime, this tool creates a container from the image and bind-mounts:
* The pipeline to the same path
* An empty `NamedTemporaryFile` to each "empty file"
* Individual pipeline files to the absolute paths defined by "mapped_files"
* A temporary directory to `/mnt/bl_tests`

Within that temporary directory, it creates:
* `docker_test.config` - the temporary config file described below
* `cli_params.json` - A JSON file of the Nextflow command-line parameters
* `test_mocks.json` - A JSON file of the test-specific method mocks

The container also gets the following environment variables:
* Everything specified by the test's "envvars"
* `BL_PIPELINE_DIR` - Path to the pipeline
* `BL_CONFIG_FILE` - The container path to `docker_test.config`
* `BL_MOCKS_FILE` - The container path to `test_mocks.json`
* `BL_CLI_PARAMS_FILE` - The container path to `cli_params.json`
* `BL_PARAMS_FILE` - The container path to the "params file" (if set)

### Temporary Config File

The generated config file looks something like this:

```groovy
import nextflow.util.SysHelper
import nextflow.util.MemoryUnit
import static org.mockito.Mockito.*
import org.mockito.MockedStatic

try (MockedStatic dummyhelper = mockStatic(
        SysHelper.class,
        CALLS_REAL_METHODS)) {
    dummyhelper
        .when(SysHelper::getAvailCpus)
        .thenReturn(16);
    dummyhelper
        .when(SysHelper::getAvailMemory)
        .thenReturn(MemoryUnit.of("31GB"));
    includeConfig "/Users/nwiltsie/src/pipeline-recalibrate-BAM/test/nftest.config"
}

System.out.println("=========SENTINEL_OUTPUT==========")
```

The two key parts here:
1. The use of mockito to mock out static Java libraries (it's mindblowing that it works) with the CPU and memory values from the test configuration.
2. Explicitly printing a sentinel value after the true config files are loaded. Our configs are littered with random print statements, so the sentinel indicates the boundary between all that junk and the structured output to follow.

### Groovy Script

Here's the main madness. This script is a chimera of [`nextflow config`](https://github.com/nextflow-io/nextflow/blob/7caffef977e0fa16177b0e7838e2b2b114c223b6/modules/nextflow/src/main/groovy/nextflow/cli/CmdConfig.groovy#L71-L114) and a portion of [`nextflow run`](https://github.com/nextflow-io/nextflow/blob/5e2ce9ed82ccbc70ec24a83e04f24b8d45855a78/modules/nextflow/src/main/groovy/nextflow/config/ConfigBuilder.groovy#L901-L906) with some major changes:

1. I don't know how to do parameter parsing in groovy, so the parameters and config files are read from the `BL_*` environment variables and injected into the `nextflow.cli.CliOptions` and `nextflow.cli.CmdRun` objects in the same way that they _would_ have been using CLI parameters.
2. The actual building of the `ConfigObject` (where all of our config code is executed) is wrapped in an [Interceptor](https://docs.groovy-lang.org/latest/html/api/groovy/lang/Interceptor.html) that gets to spy on every method call. If any of those method names are in the tests's "mocks", the actual method is ignored and the pre-computed result is returned.

At this point the `ConfigObject` is as complete as Nextflow would ever make it. However, if it were to be serialized at this point every closure would be represented like `Script1E713595D312C8FDC3D2D9EDAF11D59D$_run_closure1$_closure11@65293ca5`, which is not ideal.

To solve that I take one more step and walk over the `process` block with _another_ [Interceptor](https://docs.groovy-lang.org/latest/html/api/groovy/lang/Interceptor.html). During that process I take the following steps:
1. For top-level keys (direct children of the `process` block), cache their name (e.g. `withLabel:process_medium`) in the interceptor while traversing their children.
2. For every closure value, attempt to replace it with the following, moving onto the next if there is an issue:
	1. The result of evaluating the closure, raising an exception if it tries to [get](https://docs.groovy-lang.org/latest/html/api/groovy/util/ConfigObject.html#get(java.lang.Object)) a property named `task`.
	2. A map with keys `closure`, `1`, `2`, and `3` mapping to results of evaluating the closure, with some _serious_ skullduggery to intercept `ConfigObject.get("task")` to inject the value `[process: <saved name>, attempt: '???', cpus: '$task.cpus']` (making the closure not-so-closed). The value for `cpus` is the literal string `'$task.cpus'`, but the value for `attempt` varies:
		* For `closure`,  `attempt` is set to the literal string `'$task.attempt'`. While evaluating the closure any calls to methods named `check_limits` or `retry_updater` are mocked out for strings like `"$name(${args.join(', ')})"` (e.g. `"retry_updater(12, add, 0, $task.attempt, cpus)"`).
		* For `1`, `2`, and `3`, `attempt` is set to that integer and the closure is evaluated without any other shenanigans.
1. If we _still_ fail, it's because of a closure like `cpus = { methods.check_limits( 1 * task.attempt, 'cpus' ) }`. `1 * task.attempt` is evaluated with the static Java method `java.lang.Integer.multiply` and can't be intercepted without mockito, so at this point I just bail and return `"closure()"`.

Finally, the fully-rendered configuration is printed to stdout in Java [Properties File](https://docs.oracle.com/cd/E23095_01/Platform.93/ATGProgGuide/html/s0204propertiesfileformat01.html) format.

## Parsing and Diffing

Outside of the docker container, the groovy output is split using the sentinel value and parsed into a dictionary. A lot of junk shows up here - seemingly any variable defined without a `def` ([example](https://github.com/uclahs-cds/pipeline-Nextflow-config/blob/3ec718630ff1862377815e6c986a8b56cea1115b/config/retry/retry.config#L32), [another example](https://github.com/uclahs-cds/pipeline-Nextflow-config/blob/3ec718630ff1862377815e6c986a8b56cea1115b/config/store_object_as_json/store_object_as_json.config#L5)) bubbles up to the global namespace. For sanity I'm not parsing that `json_object` variable, but we should just add the `def` onto it.

Once parsed, I pop off any configuration-only namespaces we've defined, as they're just noise:

https://github.com/uclahs-cds/tool-Nextflow-action/blob/f8210ca71edad191f6135927837c4c76fc28644f/run-nextflow-tests/configtest.py#L224-L236

Finally, I generate a list of differences, using `re.sub` to remove any dates from the "dated_fields".

# Where to go from here

As-is, this is not an Action. However, it is _shockingly_ complicated, and before I go any further I need feedback from the lot of you. Feel free to dive into any aspect of this, but here are some specific questions:

* Is this at core a terrible idea? I don't really want to throw this away, but it will only work if the rest of you are invested in the utility of it.
* Does the top-level interface seem reasonable? My vague plan is to have the Action look for all `nf-configtest-*.json` files (or something) in the repository. That would allow each test file to be independent.

### CODE REVIEW

**Comment by nwiltsie** [comment]
> How does this work interplay, if at all, with the plan to capture all input configuration for every run?

This is definitely related work. My mental framework is as follows:

1. We write configuration and parameter files for the pipelines (**_our intended inputs_**)
2. Each pipeline's configuration process renders our inputs into variety of processes to be run and inputs for each (_**the rendered configuration**_)
3. Nextflow executes each process **(_nextflow reports, pipeline outputs_)**

There are three problems there:
* On the input side, we control the _intended_ inputs, but the pipelines actually work with the _rendered_ configuration. Something might go wrong in that translation or change inadvertently.
* On the output side, we can only infer pieces of the rendered configuration from the Nextflow reports.
* On the output side, even assuming we have the full rendered configuration, we can't back-calculate the original human inputs.

This work is partially addresses the first problem by helping us be _consistent_ in the rendering/translation process (although it doesn't say anything about it being _right_). Sufficient test coverage with this tool might convince us of the "rightness".

@j2salmingo's work with https://github.com/uclahs-cds/pipeline-Nextflow-config/pull/43 addresses the second problem by capturing the full rendered configuration with the pipeline results. 

To my knowledge, the third problem is currently unaddressed.

**Comment by sorelfitzgibbon** [comment]
Thanks, I agree it would be helpful to solve all three problems.

**Comment by kiarod** [comment]
> Is this at core a terrible idea? I don't really want to throw this away, but it will only work if the rest of you are invested in the utility of it.

On the first point, this definitely is not a terrible idea, and I don't think we can throw it away if we wanted to. This would be an awesome utility to take the guesswork out of the downstream effects of configuration changes, and would dramatically improve the quality of code reviews in a number of ways(speed, certainty, trackability of changes).

> Does the top-level interface seem reasonable? My vague plan is to have the Action look for all nf-configtest-*.json files (or something) in the repository. That would allow each test file to be independent.

In terms of top level interface, did you compose the test json by hand or did you have an automated way of constructing the json or at least a skeleton of it. Some of the components seem to be JSONified portions of the schema.yaml

Overall this is amazing work Nick! I think if this were actionified it would seriously raise the professional standard of our pipeline work. I do feel it would be worthwhile to present a demo on this work at Nextflow WG meeting so we can stop and ask you implementation level questions

**Comment by nwiltsie** [comment]
> In terms of top level interface, did you compose the test json by hand or did you have an automated way of constructing the json or at least a skeleton of it. Some of the components seem to be JSONified portions of the schema.yaml

I hand-composed a file like this:

```json
{
  "config": [
    "test/nftest.config"
  ],
  "params_file": "test/single.yaml",
  "cpus": 16,
  "memory_gb": 31,
  "empty_files": [
    "/hot/ref/reference/GRCh38-BI-20160721/Homo_sapiens_assembly38.fasta",
    "/hot/ref/tool-specific-input/GATK/GRCh38/Mills_and_1000G_gold_standard.indels.hg38.vcf.gz",
    "/hot/ref/tool-specific-input/GATK/GRCh38/Homo_sapiens_assembly38.known_indels.vcf.gz",
    "/hot/ref/tool-specific-input/GATK/GRCh38/resources_broad_hg38_v0_Homo_sapiens_assembly38.dbsnp138.vcf.gz",
    "/hot/ref/tool-specific-input/GATK/GRCh38/Biallelic/hapmap_3.3.hg38.BIALLELIC.PASS.2021-09-01.vcf.gz",
    "/hot/resource/SMC-HET/tumours/A-mini/bams/n1/output/S2.T-n1.bam"
  ],
  "mapped_files": [],
  "nf_params": {
    "output_dir": "/tmp/outputs"
  },
  "envvars": {
    "SLURM_JOB_ID": "851543"
  },
  "mocks": {
    "parse_bam_header": {
      "read_group": [
        {
          "SM": "4915723"
        }
      ]
    }
  },
  "dated_fields": [
    "params.log_output_dir",
    "report.file",
    "timeline.file",
    "trace.file"
  ],
  "expected_result": {}
}
```

So that's all of the inputs and an expected result of _nothing_. When I ran the tool it complained about all of the unexpected results, and wrote them to a new file for me (I trimmed out most of the JSON from the console output below):

```console
$ ./entry.py ~/src/pipeline-recalibrate-BAM/
docker
None
{'all_group_ids': '$(for i in `id --real --groups`; do echo -n "--group-add=$i "; done)', 'enabled': True, 'runOptions': '-u $(id -u):$(id -g) $(for i in `id --real --groups`; do echo -n "--group-add=$i "; done)', 'uid_and_gid': '-u $(id -u):$(id -g)'}
------
manifest
None
{'author': 'Yash Patel', 'description': 'Nextflow pipeline to perform Indel Realignment and Base Quality Score Recalibration', 'name': 'recalibrate-BAM', 'version': '1.0.0-rc.4'}
------
params
None
[full content trimmed]
------
params_schema
None
[full content trimmed]
------
process
None
[full content trimmed]
------
report
None
{'enabled': True, 'file': '/tmp/outputs/recalibrate-BAM-1.0.0-rc.4/TWGSAMIN000001/log-recalibrate-BAM-1.0.0-rc.4-20240215T215116Z/nextflow-log/report.html'}
------
timeline
None
{'enabled': True, 'file': '/tmp/outputs/recalibrate-BAM-1.0.0-rc.4/TWGSAMIN000001/log-recalibrate-BAM-1.0.0-rc.4-20240215T215116Z/nextflow-log/timeline.html'}
------
trace
None
{'enabled': True, 'file': '/tmp/outputs/recalibrate-BAM-1.0.0-rc.4/TWGSAMIN000001/log-recalibrate-BAM-1.0.0-rc.4-20240215T215116Z/nextflow-log/trace.txt'}
------
workDir
None
/scratch/851543
------
yaml
None
[full content trimmed]
------
Saving updated file to /Users/nwiltsie/src/tool-Nextflow-action/run-nextflow-tests/recalibrate-bam-out.json
```

So yes, that `yaml` block _is_ the JSONified version of the schema.yaml, because [this line of code](https://github.com/uclahs-cds/pipeline-Nextflow-config/blob/3ec718630ff1862377815e6c986a8b56cea1115b/config/schema/schema.config#L28) defines it without a `def`.

**Comment by nwiltsie** [comment]
(Expanding on that - [that line of code](https://github.com/uclahs-cds/pipeline-Nextflow-config/blob/3ec718630ff1862377815e6c986a8b56cea1115b/config/schema/schema.config#L28) _should_ have a `def` on it, and we _shouldn't_ include that `yaml` key in the test files. This is going to quickly turn up a lot of global variables that we didn't intend to be global.)

**Comment by zhuchcn** [comment]
This is absolutely amazing work! Have you considered incorporating this into our nf-test? It is an already published software, so we could even consider another publication of nf-test v2 in the future if more functionalities are added. Are you also expect users to write a series of test cases for different input parameters and enviroments? We can benefit from it when developing locally if this can be accessed from command line directly.

A couple of silly questions:
1. Does every closure needs to be provided in 'mocks'? What will happen if a cluster is not mocked?
2. I don't understand what entry.py is comparing against. What is a local checkout?

**Comment by nwiltsie** [comment]
> Have you considered incorporating this into our nf-test? It is an already published software, so we could even consider another publication of nf-test v2 in the future if more functionalities are added.

I've vaguely considered it - part of my goal in pushing this up now was to solicit feedback on how best to use it. I'm open to whatever makes the most sense!

> Are you also expect users to write a series of test cases for different input parameters and enviroments? We can benefit from it when developing locally if this can be accessed from command line directly.

Yes, the expectation is that we'd have a representative set of tests stored in each pipeline - different CPU counts, different inputs, etc. I agree that it should be available locally as well as in the cloud.

>  1. Does every closure needs to be provided in 'mocks'? What will happen if a cluster is not mocked?

No, not every closure needs to be mocked (in fact, as few as possible should be mocked out). Anything not mocked will be evaluated in the same way it would be in a real pipeline run.

>  2. I don't understand what entry.py is comparing against. What is a local checkout?

`entry.py` is just a small demo for this pull request. In practice this should work like NFTest, with each pipeline storing its own test files. Specifically for this pull request, to demonstrate the functionality, I bundled a test case for the [recalibrate BAM](https://github.com/uclahs-cds/pipeline-recalibrate-BAM) pipeline and wrote `entry.py` as a simple test runner. That means you can follow these steps to run the test:

```bash
#!/bin/bash

# Clone this pull request
git clone \
    --branch nwiltsie-nextflow-regression-logic \
    --recurse-submodules \
    git@github.com:uclahs-cds/tool-Nextflow-action.git

# Clone the recalibrate-BAM pipeline (with submodules)
git clone \
    --recurse-submodules \
    git@github.com:uclahs-cds/pipeline-recalibrate-BAM.git

# Run the test bundled with the pull request
./tool-Nextflow-action/run-nextflow-tests/entry.py \
    "$(readlink -f ./pipeline-recalibrate-BAM/)"
```

**Comment by nwiltsie** [comment]
https://github.com/uclahs-cds/pipeline-call-gSV/pull/118 is a perfect example of where this can be useful. @Faizal-Eeman and I created the test files from cd08fa7 and a187b4e to match pipeline-call-gSV's current `main` branch. When run against the branch from that pull request, the output is...

```console
$ ./entry.py ../../../pipelines/pipeline-call-gSV ./call-gsv-F32.json
.paramsproc_resource_params
None
{'call_gSV_Delly': {'cpus': '1', 'memory': '30 GB', 'retry_strategy': {'memory': {'operand': '2', 'strategy': 'exponential'}}}, 'call_gSV_Manta': {'cpus': '1', 'memory': '30 GB', 'retry_strategy': {'memory': {'operand': '2', 'strategy': 'exponential'}}}, 'run_validate_PipeVal': {'cpus': '1', 'memory': '1 GB'}}
------
.process.withName:call_gSV_Delly.memory
30 GB
{'1': '30 GB', '2': '60 GB', '3': '64 GB', 'closure': 'retry_updater(30 GB, exponential, 2, $task.attempt, memory)'}
------
.process.withName:call_gSV_Manta.memory
30 GB
{'1': '30 GB', '2': '60 GB', '3': '64 GB', 'closure': 'retry_updater(30 GB, exponential, 2, $task.attempt, memory)'}
------
proc_names
None
[Ljava.lang.String;@47fc9ce
------
Saving updated file to call-gsv-F32-out.json
```

So the differences are:
1. The branch newly defines the `params.proc_resource_params` variable (and my script clearly has a bug in assembling that variable name).
2. The `call_gSV_Delly` and `call_gSV_Manta` processes previously had a flat 30 GB memory allocation, and now have a progressive allocation of 30 GB, 60 GB, and 64 GB on the first three attempts (this is on an F32 node).
3. The `proc_names` variable is newly set to an opaque pointer value, indicating that it's a variable missing a `def` ([link](https://github.com/uclahs-cds/pipeline-Nextflow-config/blob/3ec718630ff1862377815e6c986a8b56cea1115b/config/retry/retry.config#L32)).

File: run-nextflow-tests/entry.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+"Quick entrypoint for this tool."
+import argparse
+import sys
+from pathlib import Path
+
+from configtest import NextflowConfigTest
+
+
+def run_pipeline_test(pipeline: Path, test_case: Path):
+    "Run the bundled pipeline-recalibrate-BAM test."
```
Comment (yashpatel6) [inline run-nextflow-tests/entry.py:11]: Docstring should be generalized?

**Review by yashpatel6 — APPROVED** [review]
Looks great! I think we can proceed with this version and make updates as necessary once we begin implementing across pipelines

**Pull request merged**